keps/sig-network/4631-loadbalancerstatus/README.md

```diff
@@ -0,0 +1,1210 @@
+  determine which tests to skip, so that any cloud provider that
+  conforms to the recommendations in this KEP will be able to run all
+  `[Feature:LoadBalancer]` tests, now and in the future, without
+  needing to have any per-provider exceptions in the test code, or
+  having to manually tell the test runner to skip any of the tests.
+
+```
+<<[UNRESOLVED auto-skipping ]>>
+
+I'm assuming that its good to have the e2e LB tests auto-self-skip, so
+that (a) we don't need per-provider skip rules in the test cases, and
+(b) we don't need per-test-case skip rules in each provider CI job.
+
+Counterarguments to this include:
+
+  1. Tests that auto-skip themselves might at some point erroneously
+     fail to distinguish "the cloud doesn't implement this feature"
+     from "the cloud is trying to implement this feature but is doing
+     it incorrectly". (The suggestions below try to minimize this by
+     having the feature detection be done in a way that is separate
+     from the actual implementation of the feature.)
+
+  2. Auto-skipping tests for newly-added features may cause cloud
+     provider maintainers to not realize that the new features exist.
+
+  3. There is a difference between automatically skipping tests for
+     features that a cloud provider _can't_ implement, vs.
+     automatically skipping tests for features that a cloud provider
+     merely doesn't implement _yet_.
+
+<<[/UNRESOLVED]>>
+```
+
+### Non-Goals
+
+- Requiring all cloud providers to implement all existing
+  load-balancer-related features.
+
+[cloud-provider-gcp]: https://github.com/kubernetes/cloud-provider-gcp/
+[cloud-provider-kind]: https://github.com/kubernetes-sigs/cloud-provider-kind/
+
+## Proposal
+
+It seems uncontroversial that the cloud provider should provide
+feedback to the user when it is in the process of provisioning a load
+balancer, or when it has conclusively failed to do so.
+
+```
+<<[UNRESOLVED split the KEP? ]>>
+
+Should we split the presumably-uncontroversial
+`LoadBalancerProvisioning` and `LoadBalancerServing` Conditions out
+into a separate KEP?
+
+<<[/UNRESOLVED]>>
+```
+
+That leaves three problems to solve.
+
+### Expected Behavior When the Cloud Provider Knows It Can't Provide a Working Load Balancer At All
+
+A cloud provider might find itself unable to provide a load balancer
+for a Service *at all* because:
+
+  - The Service uses exclusively UDP (or SCTP) ports, but the
+    cloud/load balancer only supports TCP, so a load balancer would
+    be unable to route any traffic to the Service.
+
+  - The Service is single-stack IPv6 in a cloud (or cluster) that only
+    supports IPv4 load balancers, or vice versa, so a load balancer
+    would be unable to route any traffic to the Service.
+
+  - The Service has `AllocateLoadBalancerNodePorts: false`, but the
+    cloud only supports NodePort-based load balancing.
+
+  - The Service is `ExternalTrafficPolicy: Local` but the cloud cannot
```

Comment:
> eTP=Local means what it says

lol, I thought that in the whole "what things should be traffic policy vs what things should be topology" debate (around PreferLocal) we had agreed that eTP=Local _doesn't_ mean what it says, because the actual intent of the feature is "preserve client source IP", not "route traffic in a particular way"; the routing is purely a side effect of making it possible to implement "preserve client source IP". (IOW an implementation of eTP:Local that preserves client IP while doing routing in an unexpected way would be compliant, but an implementation that routes "correctly" but loses client IP is not.)

> I think this is not really about IP preservation but has everything to do with not implementing `hCNP` or some equivalent mechanism.

What I was saying in that example was, if client IP preservation is considered a mandatory-to-implement aspect of eTP:Local, then there's no point in proxy-ish LBs implementing eTP:Local at all, and thus no reason for them to implement HCNP. (Whereas for VIP-ish LBs, there is really no good argument for not implementing HCNP.)

(If we don't think that client IP preservation is required for eTP:Local, then we should just assume all LBs will implement HCNP, and they're just buggy if they don't.)